linera-storage-service/src/server.rs

```diff
@@ -310,181 +329,365 @@ impl StoreProcessor for ServiceStoreServer {
+
+        let mut state = Vec::new();
+        let mut key = Vec::new();
+        let mut value = Vec::new();
+        let mut len_bytes = Vec::with_capacity(8);
+        while let Some(RequestWriteBatch { stream: message }) = incoming_stream.message().await? {
+            let mut message = VecDeque::from(message);
+            while !message.is_empty() || state.get(2).is_some_and(|len| 0 == *len) {
+                match state.len() {
+                    0 => {
+                        let op_kind = message.pop_front().unwrap() as usize;
+                        state.push(op_kind);
+                        state.push(0);
+                    }
+
+                    2 => {
+                        len_bytes.append(
+                            &mut message
+                                .drain(..(8 - len_bytes.len()).min(message.len()))
+                                .collect::<Vec<u8>>(),
+                        );
+                        if 8 == len_bytes.len() {
+                            state.push(usize::from_be_bytes(len_bytes.clone().try_into().unwrap()));
+                            len_bytes.clear();
                         }
-                        Some(value) => {
-                            value.extend(key_value_append.value);
+                    }
+
+                    3 if 3 == state[0] && 1 == state[1] => {
+                        value.append(
+                            &mut message
+                                .drain(..(state[2] - value.len()).min(message.len()))
+                                .collect::<Vec<_>>(),
+                        );
+                        if state[2] == value.len() {
+                            batch.put_key_value_bytes(key, value);
+                            key = Vec::new();
+                            value = Vec::new();
+                            let _ = state.pop();
+                            let _ = state.pop();
+                            let _ = state.pop();
                         }
                     }
-                    if key_value_append.last {
-                        let value = pending_big_puts.remove(&key_value_append.key).unwrap();
-                        batch.put_key_value_bytes(key_value_append.key, value);
+
+                    3 => {
+                        key.append(
+                            &mut message
+                                .drain(..(state[2] - key.len()).min(message.len()))
+                                .collect::<Vec<_>>(),
+                        );
+                        match (state[0], state[2] == key.len()) {
+                            (1, true) => {
+                                batch.delete_key(key);
+                                key = Vec::new();
+                                let _ = state.pop();
+                                let _ = state.pop();
+                                let _ = state.pop();
```

Comment:
It is a little unfortunate to have the 3 pops. But the alternativer with `split_off` or `drain` is not much simpler.